### ISSUE TITLE
CASSCF kernel function impurity leads to persistent _max_stepsize reduction in repeated kernel calls

### ISSUE THREAD

**Issue by MatthewRHermes** [pyscf/pyscf#1762]
The following:
```
from pyscf import gto, scf, mcscf, lib
xyz='''C 0 0 0,
       O 1 0 0'''
mol = gto.M (atom=xyz, basis='cc-pvdz', symmetry=False, verbose=lib.logger.DEBUG, output='CO.log')
mf = scf.RHF (mol).run ()
mo = mf.mo_coeff.copy ()
mc = mcscf.CASSCF (mf, 8, 10).run () # default convergence requested
print ("mc.converged =", mc.converged, "; mc.e_tot =", mc.e_tot, '; mc._max_stepsize =', mc._max_stepsize)
mc.run (mo, conv_tol_grad=1e-16) # impossibly tight convergence requested
print ("mc.converged =", mc.converged, "; mc.e_tot =", mc.e_tot, '; mc._max_stepsize =', mc._max_stepsize)
mc.run (mo, conv_tol_grad=None) # default convergence requested
print ("mc.converged =", mc.converged, "; mc.e_tot =", mc.e_tot, '; mc._max_stepsize =', mc._max_stepsize)
```
results in the following:
```
overwrite output file: CO.log
mc.converged = True ; mc.e_tot = -112.81902947003246 ; mc._max_stepsize = 0.02
mc.converged = False ; mc.e_tot = -112.81902947761458 ; mc._max_stepsize = 6.565139347890731e-25
mc.converged = False ; mc.e_tot = -112.74869243452127 ; mc._max_stepsize = 2.742069470056062e-39
```
That is, the third calculation, which should be identical to the first, is unable to converge, due to the effects of the `_max_stepsize` member variable which is gradually constricted to essentially zero over the course of the futile second calculation. The issue is resolved by resetting `_max_stepsize` manually:
```
from pyscf import gto, scf, mcscf, lib
xyz='''C 0 0 0,
       O 1 0 0'''
mol = gto.M (atom=xyz, basis='cc-pvdz', symmetry=False, verbose=lib.logger.DEBUG, output='CO.log')
mf = scf.RHF (mol).run ()
mo = mf.mo_coeff.copy ()
mc = mcscf.CASSCF (mf, 8, 10)
print ("The initial value of mc._max_stepsize is", mc._max_stepsize)
mc.run () # default convergence requested
print ("mc.converged =", mc.converged, "; mc.e_tot =", mc.e_tot, '; mc._max_stepsize =', mc._max_stepsize)
mc.run (mo, conv_tol_grad=1e-16) # impossibly tight convergence requested
print ("mc.converged =", mc.converged, "; mc.e_tot =", mc.e_tot, '; mc._max_stepsize =', mc._max_stepsize)
mc.run (mo, conv_tol_grad=None, _max_stepsize=None) # default convergence requested
print ("mc.converged =", mc.converged, "; mc.e_tot =", mc.e_tot, '; mc._max_stepsize =', mc._max_stepsize)
```
results in expected behavior:
```
overwrite output file: CO.log
The initial value of mc._max_stepsize is None
mc.converged = True ; mc.e_tot = -112.81902947436106 ; mc._max_stepsize = 0.02
mc.converged = False ; mc.e_tot = -112.81902947761158 ; mc._max_stepsize = 6.565139347890731e-25
mc.converged = True ; mc.e_tot = -112.81902947436123 ; mc._max_stepsize = 0.02
```
This issue might severely affect potential energy scans, or even geometry optimizations if `_max_stepsize` gets altered by an energy calculation that reports successful convergence, in a way that is extremely subtle and difficult to diagnose. I suspect it wasn't intended for `_max_stepsize` to carry over from kernel call to kernel call like this. In fact, the mc1step.py version of the function that controls `_max_stepsize` is strangely different from the "sibling" function in umc1step.py:
 ```
> grep -A11 "def max_stepsize_scheduler(self, envs)" *.py
mc1step.py:    def max_stepsize_scheduler(self, envs):
mc1step.py-        if not WITH_STEPSIZE_SCHEDULER:
mc1step.py-            return self.max_stepsize
mc1step.py-
mc1step.py-        if self._max_stepsize is None:
mc1step.py-            self._max_stepsize = self.max_stepsize
mc1step.py-        if envs['de'] > -self.conv_tol:  # Avoid total energy increasing
mc1step.py-            self._max_stepsize *= .3
mc1step.py-            logger.debug(self, 'set max_stepsize to %g', self._max_stepsize)
mc1step.py-        else:
mc1step.py-            self._max_stepsize = (self.max_stepsize*self._max_stepsize)**.5
mc1step.py-        return self._max_stepsize
--
umc1step.py:    def max_stepsize_scheduler(self, envs):
umc1step.py-        if self._max_stepsize is None:
umc1step.py-            self._max_stepsize = self.max_stepsize
umc1step.py-        if envs['de'] > self.conv_tol:  # Avoid total energy increasing
umc1step.py-            self._max_stepsize *= .5
umc1step.py-            logger.debug(self, 'set max_stepsize to %g', self._max_stepsize)
umc1step.py-        else:
umc1step.py-            self._max_stepsize = numpy.sqrt(self.max_stepsize*self.max_stepsize)
umc1step.py-        return self._max_stepsize
 ```
 is the last leading underscore in the penultimate line of the mc1step.py version a typo?

**Comment by MatthewRHermes** [comment]
If I change the tightness of the calculation using `conv_tol` (as most users would), rather than `conv_tol_grad`, then there is still a problem, but the precise nature of the problem is inconsistent. The outcome of the third calculation seems to depend on just how close to zero `mc._max_stepsize` gets over the course of the futile second calculation, which seems to depend on rounding error in some way.

**Comment by MatthewRHermes** [comment]
It's also strange to me that in the mc1step.py version, the condition for constricting _max_stepsize uses the negative sign (so that dE = 0 triggers a constriction), but in the mc1step.py version, it uses a positive sign (so that only energy increasing triggers a constriction). I guess one of these was "fiddled with" for optimization purposes without changing the other one?

**Comment by sunqm** [comment]
> It's also strange to me that in the mc1step.py version, the condition for constricting _max_stepsize uses the negative sign (so that dE = 0 triggers a constriction), but in the mc1step.py version, it uses a positive sign (so that only energy increasing triggers a constriction). I guess one of these was "fiddled with" for optimization purposes without changing the other one?

Do you mean the minus sign in the condition `if envs['de'] > -self.conv_tol:` around L1240 of mc1step ? I guess the sign has negligible impacts. Of cause it is better to have the sign consistency in mc1step.py and umc1step.py . Would you mind to correct them in your PR?

**Comment by MatthewRHermes** [comment]
Sure, the sign probably has negligible impacts, and I can change it in my PR. Regarding the other inconsistency, though, the one between mc1step.py#L1243 and umc1step.py#L813, could you tell me which behavior is intended? I suppose it's probably the former (i.e., gradually release the constricted _max_stepsize instead of abruptly resetting it), but I'm not 100% sure.

**Comment by MatthewRHermes** [comment]
Fixed by #1763
